torch/distributed/_spmd/data_parallel.py

```diff
@@ -0,0 +1,795 @@
+        reduction_over_batch: bool = False,
+    ):
+        super().__init__(startegy_list)
+        self.node_type = node_type
+        self.reduction_over_batch = reduction_over_batch
+
+    def __str__(self) -> str:
+        return f"type: {self.node_type}, {super().__str__()}"
+
+
+@contextmanager
+def gradients_tagging(params: Dict[str, torch.Tensor]):
+    """
+    This is a helper function that tags the gradient of the parameters
+    with a special tag, so that we can identify them during SPMD expansion.
+
+    It's safe to trace those hooks and we would remove those nodes later.
+    """
+
+    tagging_hooks = []
+    try:
+        for p in params.values():
+            h = p.register_hook(lambda grad: torch.ops._spmd.tag_grad(grad))
+            tagging_hooks.append(h)
+        yield
+    finally:
+        # remove those hooks after tracing
+        for h in tagging_hooks:
+            h.remove()
+
+
+class BatchDimAnalyzer(object):
+    """
+    This class is used to analyze the batch dimension of each tensor/node in the
+    graph. We need to know the batch dimension of each tensor/node so that we know
+    exactly the sharding layout of intermediate tensors.
+
+    We possibly should use symbolic shapes to track the batch dimension, but this
+    needs to happen with dynamo, as dynamo is the only place we can mark a single
+    dimension. For now, we just use the batch dimension of the first input tensor
+    as the hint to track the batch dimension of all tensors/nodes in the graph.
+    """
+
+    def __init__(self, batch_dim: int = 0) -> None:
+        self.batch_dim = batch_dim
+
+        if batch_dim != 0:
+            # TODO: see if this make sense or not
+            raise RuntimeError("Data Parallel only supports batch dim on dimension 0!")
+
+        self.batch_dim_map: Dict[fx.Node, int] = {}
+        # batch dim size is used to track the batch dim size of the input tensor
+        self.batch_dim_size = -1
+
+    def init_batch_dim_size(self, batch_dim_size: int) -> None:
+        """
+        initialize batch dim size base on the first input batch size
+        """
+        if self.batch_dim_size != -1 and self.batch_dim_size != batch_dim_size:
+            raise RuntimeError(
+                f"batch dim size is already initialized! "
+                f"Found new batch size: {batch_dim_size} not "
+                f"matching existing batch dim size: {self.batch_dim_size}!"
+            )
+        self.batch_dim_size = batch_dim_size
+
+    def set_batch_dim(self, node: fx.Node, batch_dim: int) -> None:
+        self.batch_dim_map[node] = batch_dim
+
+    def get_batch_dim(self, node: fx.Node) -> int:
+        if node not in self.batch_dim_map:
+            raise RuntimeError(f"batch dim analysis failed on node: {node}!")
+        return self.batch_dim_map[node]
+
+    def get_batch_dim_shard_spec(
+        self, node: fx.Node, mesh: DeviceMesh, input_full_reduction: bool = False
+    ) -> Tuple[DTensorSpec, bool]:
```

Comment:
It feels like this can be factored into two functions: one function that computes the batch dim, and then another which computes the sharding spec from the batch dim.